### ISSUE TITLE
Adversarial attack on Detector

### ISSUE THREAD

**Issue by AlexeyAB** [AlexeyAB/darknet#5105]
Try to find Dog, Car, Bicycle on this image: [drawn.zip](https://github.com/AlexeyAB/darknet/files/4377453/drawn.zip)

By using Yolov3-model
`./darknet detector test cfg/coco.data yolov3.cfg yolov3.weights dog.jpg`

![image](https://user-images.githubusercontent.com/4096485/77472580-375e6580-6e25-11ea-8179-ae95d674e844.png)

----

How to use:

1. Build Darknet with GPU=1 CUDNN=1 OPENCV=1
2. Run command: 
`./darknet detector draw cfg/coco.data yolov3.cfg yolov3.weights -thresh 0.25 dog.jpg`

3. Move trackbar:
    * iterations=500
    * learning_rate exp = 20
    * class_id = 15

4. Select any area on the image by Left-mouse-button

![image](https://user-images.githubusercontent.com/4096485/77472553-2ca3d080-6e25-11ea-93a9-c3da2c864c5f.png)

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

I could detect the dog with these parameters. Can you explain more about this feature?


![ddd](https://user-images.githubusercontent.com/5433346/77751593-a0bbbf80-7043-11ea-8f63-eef8a01bb264.jpg)

**Comment by AlexeyAB** [comment]
@zpmmehrdad 
> Can you explain more about this feature?

* Regular training - network during training changes their weights to required detect objects
* Adversarial attack - network during training changes initial image to required detect objects, so you can make this neural network detect the objects you need, and not detect unnecessary

I improved adversarial attack and training. Download the latest Darknet code.
Now use default settings.

![image](https://user-images.githubusercontent.com/4096485/77761188-b41d5980-7048-11ea-96e9-57904f10660f.png)

**Comment by HagegeR** [comment]
this is a kind of transfer learning or fine tuning? won't it affect global performance (improving one class while degrading others) ?

**Comment by AlexeyAB** [comment]
@HagegeR This is a utility for this to fool the detector. It does't train the model, it train the image. 
To train the model use this: https://github.com/AlexeyAB/darknet/issues/5117

**Comment by sisrfeng** [comment]
> @zpmmehrdad
> 
> > Can you explain more about this feature?
> 
> * Regular training - network during training changes their weights to required detect objects
> * Adversarial attack - network during training changes initial image to required detect objects, so you can make this neural network detect the objects you need, and not detect unnecessary
> 
> I improved adversarial attack and training. Download the latest Darknet code.
> Now use default settings.
> 
> ![image](https://user-images.githubusercontent.com/4096485/77761188-b41d5980-7048-11ea-96e9-57904f10660f.png)

Should "to required detect objects"  be "to detect required objects" ?
 By adding  some noise on the dog to make it look like a cat (to machine but not to human), then training the model to detect the modified dog ,  "so you can make this neural network detect the objects you need (dog)" , right? 
Similar to this?
![image](https://user-images.githubusercontent.com/53520949/80596342-ced45b00-8a58-11ea-9579-11a877f72689.png)

**Comment by AlexeyAB** [comment]
@sisrfeng Yes, but since we have access to the structure+weights of neural network (not just black-box), so we can do such attack more efficient (with less changes - like side effects of image-compression). 
But we can also teach it to defend herself against such attacks.
